File: 0097-Interleaving String/0097-Interleaving String.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int m = s1.size();
        int n = s2.size();
        if (s3.size() != m + n)
        {
            return false;
        }
        
        vector<vector<bool>> dp(1 + m, vector<bool>(1 + n));
        
        dp[0][0] = true;
        for (int i = 1; i <= m && dp[i - 1][0]; ++i)
        {
            dp[i][0] = (s1[i - 1] == s3[i - 1]);
        }
        
        for (int j = 1; j <= n && dp[0][j - 1]; ++j)
        {
            dp[0][j] = (s2[j - 1] == s3[j - 1]);
        }
        
        for (int i = 1; i <= m; ++i)
        {
            for (int j = 1; j <= n; ++j)
            {
                dp[i][j] = (s1[i - 1] == s3[i + j - 1]) && dp[i - 1][j] ||
                           (s2[j - 1] == s3[i + j - 1]) && dp[i][j - 1];
            }
        }
        
        return dp[m][n];
    }
};

// Optimize space to O(n)
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int m = s1.size();
        int n = s2.size();
        if (s3.size() != m + n)
        {
            return false;
        }
        
        vector<bool> dp(1 + n);
        dp[0] = true;
        for (int j = 1; j <= n && dp[j - 1]; ++j)
        {
            dp[j] = (s2[j - 1] == s3[j - 1]);
        }
        for (int i = 1; i <= m; ++i)
        {
            vector<bool> temp(1 + n);
            temp[0] = dp[0] && (s1[i - 1] == s3[i - 1]);
            for (int j = 1; j <= n; ++j)
            {
                temp[j] = (s1[i - 1] == s3[i + j - 1]) && dp[j] ||
                           (s2[j - 1] == s3[i + j - 1]) && temp[j - 1];
            }
            dp = temp;
        }
        
        return dp[n];
    }
};

File: 0001-0100/0097-Interleaving String/0097-Interleaving String.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        m, n = len(s1), len(s2)
        if len(s3) != m + n:
            return False
        dp = [[False] * (1 + n) for _ in range(1 + m)]
        dp[0][0] = True
        for i in range(1, 1 + m):
            if s1[i - 1] == s3[i - 1]:
                dp[i][0] = True
            else:
                break
                
        for j in range(1, 1 + n):
            if s2[j - 1] == s3[j - 1]:
                dp[0][j] = True
            else:
                break
                
        for i in range(1, 1 + m):
            for j in range(1, 1 + n):
                if (s1[i - 1] == s3[i + j - 1] and dp[i - 1][j] or
                    s2[j - 1] == s3[i + j - 1] and dp[i][j - 1]):
                    dp[i][j] = True
        return dp[-1][-1]


# Optimize space
class Solution2:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s3) != len(s1) + len(s2):
            return False
        dp = [False] * (1 + len(s2))
        dp[0] = True
        for j in range(1, len(dp)):
            if s2[j - 1] == s3[j - 1]:
                dp[j] = True
            else:
                break
                
        for i in range(1, len(s1) + 1):
            temp = [False] * (1 + len(s2))
            temp[0] = dp[0] and (s1[i - 1] == s3[i - 1])
            for j in range(1, len(s2) + 1):
                temp[j] = (((s1[i - 1] == s3[i + j - 1]) and dp[j]) or
                    ((s2[j - 1] == s3[i + j - 1]) and temp[j - 1]))
            dp = temp
        return dp[-1]
